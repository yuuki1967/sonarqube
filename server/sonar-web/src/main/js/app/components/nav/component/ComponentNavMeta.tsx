/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { connect } from 'react-redux';
import ComponentNavWarnings from './ComponentNavWarnings';
import BranchStatus from '../../../../components/common/BranchStatus';
import DateTimeFormatter from '../../../../components/intl/DateTimeFormatter';
import Favorite from '../../../../components/controls/Favorite';
import HomePageSelect from '../../../../components/controls/HomePageSelect';
import Tooltip from '../../../../components/controls/Tooltip';
import DetachIcon from '../../../../components/icons-components/DetachIcon';
import {
  isShortLivingBranch,
  isLongLivingBranch,
  isMainBranch,
  isPullRequest
} from '../../../../helpers/branches';
import { translate } from '../../../../helpers/l10n';
import { isLoggedIn } from '../../../../helpers/users';
import { getCurrentUser, Store } from '../../../../store/rootReducer';

interface StateProps {
  currentUser: T.CurrentUser;
}

interface Props extends StateProps {
  branchLike?: T.BranchLike;
  component: T.Component;
  warnings: string[];
}

export function ComponentNavMeta({ branchLike, component, currentUser, warnings }: Props) {
  const mainBranch = !branchLike || isMainBranch(branchLike);
  const longBranch = isLongLivingBranch(branchLike);
  const currentPage = getCurrentPage(component, branchLike);
  const displayVersion = component.version !== undefined && (mainBranch || longBranch);

  return (
    <div className="navbar-context-meta">
      {warnings.length > 0 && <ComponentNavWarnings warnings={warnings} />}
      {component.analysisDate && (
        <div className="spacer-left text-ellipsis">
          <DateTimeFormatter date={component.analysisDate} />
        </div>
      )}
      {displayVersion && (
        <Tooltip mouseEnterDelay={0.5} overlay={`${translate('version')} ${component.version}`}>
          <div className="spacer-left text-limited">
            {translate('version')} {component.version}
          </div>
        </Tooltip>
      )}
      {isLoggedIn(currentUser) && (
        <div className="navbar-context-meta-secondary">
          {mainBranch && (
            <Favorite
              component={component.key}
              favorite={Boolean(component.isFavorite)}
              qualifier={component.qualifier}
            />
          )}
          {(mainBranch || longBranch) &&
            currentPage !== undefined && (
              <HomePageSelect className="spacer-left" currentPage={currentPage} />
            )}
        </div>
      )}
      {(isShortLivingBranch(branchLike) || isPullRequest(branchLike)) && (
        <div className="navbar-context-meta-secondary display-inline-flex-center">
          {isPullRequest(branchLike) &&
            branchLike.url !== undefined && (
              <a
                className="display-inline-flex-center big-spacer-right"
                href={branchLike.url}
                rel="noopener noreferrer"
                target="_blank">
                {translate('branches.see_the_pr')}
                <DetachIcon className="little-spacer-left" size={12} />
              </a>
            )}
          <BranchStatus branchLike={branchLike} />
        </div>
      )}
    </div>
  );
}

function getCurrentPage(component: T.Component, branchLike: T.BranchLike | undefined) {
  let currentPage: T.HomePage | undefined;
  if (component.qualifier === 'VW' || component.qualifier === 'SVW') {
    currentPage = { type: 'PORTFOLIO', component: component.key };
  } else if (component.qualifier === 'APP') {
    const branch = isLongLivingBranch(branchLike) ? branchLike.name : undefined;
    currentPage = { type: 'APPLICATION', component: component.key, branch };
  } else if (component.qualifier === 'TRK') {
    // when home page is set to the default branch of a project, its name is returned as `undefined`
    const branch = isLongLivingBranch(branchLike) ? branchLike.name : undefined;
    currentPage = { type: 'PROJECT', component: component.key, branch };
  }
  return currentPage;
}

const mapStateToProps = (state: Store): StateProps => ({
  currentUser: getCurrentUser(state)
});

export default connect(mapStateToProps)(ComponentNavMeta);
