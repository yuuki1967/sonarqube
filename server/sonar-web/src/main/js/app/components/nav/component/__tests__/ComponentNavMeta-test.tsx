/*
 * SonarQube
 * Copyright (C) 2009-2019 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { shallow } from 'enzyme';
import { ComponentNavMeta } from '../ComponentNavMeta';
import {
  mockShortLivingBranch,
  mockComponent,
  mockLongLivingBranch,
  mockPullRequest
} from '../../../../../helpers/testMocks';

it('renders status of short-living branch', () => {
  expect(shallowRender()).toMatchSnapshot();
});

it('renders meta for long-living branch', () => {
  expect(shallowRender({ branchLike: mockLongLivingBranch() })).toMatchSnapshot();
});

it('renders meta for pull request', () => {
  expect(
    shallowRender({
      branchLike: mockPullRequest({
        url: 'https://example.com/pull/1234'
      })
    })
  ).toMatchSnapshot();
});

function shallowRender(props = {}) {
  return shallow(
    <ComponentNavMeta
      branchLike={mockShortLivingBranch()}
      component={mockComponent({ analysisDate: '2017-01-02T00:00:00.000Z', version: '0.0.1' })}
      currentUser={{
        isLoggedIn: false
      }}
      warnings={[]}
      {...props}
    />
  );
}
