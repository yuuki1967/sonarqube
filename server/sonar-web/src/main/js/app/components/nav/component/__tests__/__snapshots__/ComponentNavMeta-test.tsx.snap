// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders meta for long-living branch 1`] = `
<div
  className="navbar-context-meta"
>
  <div
    className="spacer-left text-ellipsis"
  >
    <DateTimeFormatter
      date="2017-01-02T00:00:00.000Z"
    />
  </div>
  <Tooltip
    mouseEnterDelay={0.5}
    overlay="version 0.0.1"
  >
    <div
      className="spacer-left text-limited"
    >
      version
       
      0.0.1
    </div>
  </Tooltip>
</div>
`;

exports[`renders meta for pull request 1`] = `
<div
  className="navbar-context-meta"
>
  <div
    className="spacer-left text-ellipsis"
  >
    <DateTimeFormatter
      date="2017-01-02T00:00:00.000Z"
    />
  </div>
  <div
    className="navbar-context-meta-secondary display-inline-flex-center"
  >
    <a
      className="display-inline-flex-center big-spacer-right"
      href="https://example.com/pull/1234"
      rel="noopener noreferrer"
      target="_blank"
    >
      branches.see_the_pr
      <DetachIcon
        className="little-spacer-left"
        size={12}
      />
    </a>
    <BranchStatus
      branchLike={
        Object {
          "analysisDate": "2018-01-01",
          "base": "master",
          "branch": "feature/foo/bar",
          "key": "1001",
          "title": "Foo Bar feature",
          "url": "https://example.com/pull/1234",
        }
      }
    />
  </div>
</div>
`;

exports[`renders status of short-living branch 1`] = `
<div
  className="navbar-context-meta"
>
  <div
    className="spacer-left text-ellipsis"
  >
    <DateTimeFormatter
      date="2017-01-02T00:00:00.000Z"
    />
  </div>
  <div
    className="navbar-context-meta-secondary display-inline-flex-center"
  >
    <BranchStatus
      branchLike={
        Object {
          "analysisDate": "2018-01-01",
          "isMain": false,
          "mergeBranch": "master",
          "name": "release-1.0",
          "type": "SHORT",
        }
      }
    />
  </div>
</div>
`;
